File: CS320/HW3/Tree.py
# -*- coding: utf-8 -*-
"""
Created on Sun Oct  9 17:12:32 2016

@author: xt
"""
# Teng Xu xt@bu.edu
# HW3

# Answer to 4(c):
# Advantages of higher-order functions:
# 1. Higher order function is applicable to many different problems.
# 2. Higher order function has a relativly fixed syntax so we will
# not be confused during writing the code and it also provides us with a
# idea or a hint on how to slove the problem.


class Tree:
    def __init__(self, a, left, right):
        self.value = a
        self.left = left
        self.right = right
    
    def __eq__(self,other):
        return self.List2() == other

    
    def leafCount(self):
        if(self.value == None):
            return 1
        elif(self.left != None) and (self.right != None):
            return self.left.leafCount() + self.right.leafCount()
        elif(self.left == None) and (self.right != None):
            return 1 + self.right.leafCount()
        elif(self.left != None) and (self.right == None):
            return 1 + self.left.leafCount()
        else:
            return 2
    
    def nodeCount(self):
        if(self.value == None):
            return 0
        elif(self.left != None) and (self.right != None):
            return 1 + self.left.nodeCount() + self.right.nodeCount()
        elif(self.left == None) and (self.right != None):
            return 1 + self.right.nodeCount()
        elif(self.left != None) and (self.right == None):
            return 1 + self.left.nodeCount()
        else:
            return 1
        
    def height(self):
        if(self.value == None):
            return 0
        elif(self.left != None) and (self.right != None):
            maxvalue = max(self.left.height(), self.right.height())
            return 1 + maxvalue
        elif(self.left == None) and (self.right != None):
            return 1 + self.right.nodeCount()
        elif(self.left != None) and (self.right == None):
            return 1 + self.left.nodeCount()
        else:
            return 1           
    
    def perfect(self):
        if(self.value == None):
            return True
        elif(self.left != None) and (self.right != None):
            if (self.left.height() == self.right.height()):
                return (self.left.perfect() and self.right.perfect())
        elif(self.left == None) and (self.right == None):
            return True
        else:
            return False
            
    def degenerate(self):
        if(self.value == None):
            return True   
        elif(self.left != None) and (self.right != None):
            return self.left.degenerate() and self.right.degenerate()
        elif(self.left == None) and (self.right != None):
            return self.right.degenerate()
        elif(self.left != None) and (self.right == None):
            return self.left.degenerate()
        else:
            return True
    
    def List(self):
        if(self.value == None or self == None):
            return []
        elif(self.degenerate == False):
            return []
        elif(self.left == None) and (self.right != None):
            return [self.value] + self.right.List()
        elif(self.left != None) and (self.right == None):
            return [self.value] + self.left.List()
        else:
            return [self.value]   
            
# get a list representation of the tree, helper function for equal 
    def List2(self):
        if(self.value == None):
            return [0,0,0]
        elif(self.left == None) and (self.right != None):
            return [self.value] + [0] + self.right.List2()
        elif(self.left != None) and (self.right == None):
            return [self.value] + self.left.List2() + [0]
        else:
            return [self.value] + [0] + [0]
        

File: CS320/HW3/Tree.hs
-- Teng Xu xt@bu.edu
-- HW3

--Leaf :: Tree a
--Node :: a -> Tree a -> Tree a -> Tree a
data Tree a = Leaf
            | Node a (Tree a) (Tree a)
              deriving (Show, Eq)

--mapT :: (a -> b) -> Tree a -> Tree b
mapT f Leaf = Leaf
mapT f (Node a left right) = Node (f a) (mapT f left) (mapT f right)

--foldT :: (a -> b -> b -> b) -> b -> Tree a -a> b
foldT f b Leaf = b
foldT f b (Node a left right) = f a (foldT f b left) (foldT f b right)

--leafCount :: Tree a -> Integer
leafCount Leaf = 1
leafCount (Node a left right) = foldT sum 1 (Node a left right)
                                    where
                                    sum x y z = y + z

--nodeCount :: Tree a -> Integer
nodeCount Leaf = 0
nodeCount (Node a left right) = foldT sum2 0 (Node a left right)
                                    where
                                    sum2 x y z = 1 + y + z

--height :: Tree a -> Integer
height Leaf = 0
height (Node a left right) = foldT sum3 0 (Node a left right)
                                    where
                                    sum3 x y z
                                        | y > z = 1 + y
                                        | otherwise = 1 + z

--perfect :: Tree a -> Bool
perfect Leaf = True
perfect (Node a left right)
    | ((height left) == (height right)) = (perfect left) && (perfect right)
    | otherwise = False

--degenerate :: Tree a -> Bool
degenerate Leaf = True
degenerate (Node a left right)
    | (left == Leaf || right == Leaf) = (degenerate left) && (degenerate right)
    | otherwise = False

--list :: Tree a -> Maybe [a]
list Leaf = []
list (Node a left right)
    | (degenerate (Node a left right) == False) = []
    | otherwise = foldT f [] (Node a left right)
                  where
                  f x y z = [x] ++ y ++ z
